Replace metricsmap-bpf-prom-sync with Prometheus Collector pattern

carnerito opened the pull request:
Previously, the 'metricsmap-bpf-prom-sync' controller was responsible for periodically collecting cilium_datapath drop
and forward metrics.
This PR introduces the metricsmapCollector, which implements the Prometheus Collector interface within the
metricsmap package. As a result, the aforementioned controller has been removed.
The metricsmapCollector is registered within the global Prometheus registry during the initialization of metricsmap.
The metrics.Register function has been modified to propagate errors from the registry.Register function instead of
simply overriding it with nil.
The metricsmapCollector comprises two metrics maps: forwardedMetricsMap and droppedMetricsMap.
These maps are populated within a callback function passed to the IterateWithCallback function.
This approach serves two primary purposes:
1. Separation of Map Iteration and Metric Update: By separating the iteration over the BPF map and the updating of
    Prometheus metrics, the implementation ensures that no partial metrics are exposed in case of map iteration failure.
2. Normalization of exposed Metrics: Unlike the statement in the bpf/lib/metrics.h comments, which suggests exposing
    only one reason label for forwarded metrics, the eBPF code exposes multiple reasons. Through testing, it was found
    that reasons 0 and 3 were exposed which triggered this error in prometheus client:
    https://github.com/prometheus/client_golang/issues/242

Furthermore, cilium command has an indirect dependency on metricsmap package. To prevent metrics map initialization
from happening each time cilium command is executed, metricsmap is converted to hive Cell and injected in cilium daemon
Infrastructure module. For details see this comment: https://github.com/cilium/cilium/pull/27370#issuecomment-1710805663.

Fixes: #27058

Signed-off-by: Boris Petrovic <carnerito.b@gmail.com>

zacharysarah commented on Documentation/bpf/toolchain.rst:
```diff
@@ -144,7 +144,8 @@ In order to run through all BPF selftests, the following command is needed:
 
-If you see any failures, please contact us on Slack with the full test output.
+If you see any failures, please contact us on `Cilium Slack`_ with the full
+test output.
```
Who is "us"? I know this is inherited language, but it's an opportunity to be specific and helpful rather than remaining vague.

carnerito replied:
Not part of this PR.

zacharysarah commented on Documentation/community/governance/commit_access.rst:
```diff
@@ -130,11 +130,13 @@ The Maintainers may also choose to delegate working with the CNCF to non-Maintai
-committers Slack channel. If the reported Code of Conduct violator is a Maintainer, the 
-Maintainers will instead designate two Maintainers to work with the
-`CNCF CoC Committee <https://www.cncf.io/conduct/procedures/>`__.  
+`Code of Conduct`_ violations by community members will be discussed and
```
Use present tense
```suggestion
`Code of Conduct`_ violations by community members are discussed and
```

carnerito replied:
Not part of this PR.

zacharysarah commented on Documentation/community/governance/commit_access.rst:
```diff
@@ -349,8 +351,8 @@ decide an issue. If the maintainers themselves cannot decide an issue, the
-Votes are done in the Slack channel #committers using Slack polls. A failure to
-vote is an implicit abstention.
+Votes are done in the Slack channel ``#committers`` using Slack polls. A
+failure to vote is an implicit abstention.
```
I think we can be gentler.
```suggestion
Votes are done in the Slack channel ``#committers`` using Slack polls.
Declining to vote is an implicit abstention.
```

carnerito replied:
Not part of this PR.

zacharysarah commented on Documentation/community/roadmap.rst:
```diff
@@ -171,15 +171,18 @@ production.
+working on modularizing the codebase and going from a tightly coupled design
+(one large initialization and configuration) to a more loosely coupled design
+of mostly self-contained modules. This will make Cilium internals easier to
+comprehend, test and extend.
```
Avoid "we", use plain, clear language, specific is better than general, future tense is OK here
```suggestion
As the project grows in complexity, it is increasingly important
to be able to divide the project into more manageable chunks. To achieve this, contributors
are modularizing the codebase and going from a tightly coupled design
(one large initialization and configuration) to a more loosely coupled design
of mostly self-contained modules. This will make Cilium internals easier to
comprehend, test, and extend.
```

carnerito replied:
Not part of this PR.

zacharysarah commented on Documentation/community/roadmap.rst:
```diff
@@ -171,15 +171,18 @@ production.
+Contributions in this area are very welcome. To get started, take a look at the
+:ref:`guide-to-the-hive` documentation and the issues referenced from
+`modularization meta issue <modularization-issue_>`_. If you have any questions
+or ideas please join us on the ``#sig-modularization`` channel on `Cilium
```
Use plain, clear language (avoid "please")
```suggestion
or ideas, join us on the ``#sig-modularization`` channel on `Cilium
```

carnerito replied:
Not part of this PR.

zacharysarah commented on Documentation/configuration/argocd-issues.rst:
```diff
@@ -83,13 +83,19 @@ Solution
-If you have an issue with Argo CD that's not outlined above, check this `list of Argo CD related issues on GitHub <https://github.com/cilium/cilium/issues?q=is%3Aissue+argocd>`__.
-If you can't find an issue that relates to yours, create one and/or seek help on the :term:`Slack channel`.
+
+    When using ``helm template``, it is highly recommended you set
```
Use clear, plain language
```suggestion
    When using ``helm template``, set
```

carnerito replied:
Not part of this PR.

zacharysarah commented on Documentation/configuration/argocd-issues.rst:
```diff
@@ -83,13 +83,19 @@ Solution
+    By specifying ``--api-versions=monitoring.coreos.com/v1`` you should be
+    able to pass validation with ``helm template``.
+
+If you have an issue with Argo CD that's not outlined above, check this `list
```
accessibility nit
```suggestion
If you have an issue with Argo CD that's not outlined above, review this `list
```

carnerito replied:
Not part of this PR.

zacharysarah commented on Documentation/glossary.rst:
```diff
@@ -12,8 +12,8 @@ Glossary
 them below with short descriptions. If you need more information, please ask us
-on `Slack <https://cilium.herokuapp.com>`_. Feel free to extend this document
-with words you expected to see here.
+on `Cilium Slack`_. Feel free to extend this document with words you expected
```
wording nit
```suggestion
on `Cilium Slack`_. Feel free to extend this document with terms you expected
```

carnerito replied:
Not part of this PR.

zacharysarah commented on Documentation/installation/cni-chaining-generic-veth.rst:
```diff
@@ -34,7 +34,8 @@ Validate that the current CNI plugin is using veth
 ``generic-veth`` plugin is not suitable. In that case, a full CNI chaining
 plugin is required which understands the device model of the underlying plugin.
-Writing such a plugin is trivial, contact us on :ref:`slack` for more details.
+Writing such a plugin is trivial, contact us on `Cilium Slack`_ for more
```
Who is "us"? I recognize this is inherited language, but it's a chance to be specific and helpful rather than continuing to be vague.

carnerito replied:
Not part of this PR.

zacharysarah commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -137,47 +137,134 @@ func (vs Values) Bytes() uint64 {
+
+	// eBPF code seems to expose multiple reasons for forwarded metrics
+	// as opposed to what is stated in bpf/lib/metrics.h comments.
+	// IterateWithCallback will iterate through BPF map for each reason and direction.
```
Verb tense nit
```suggestion
	// IterateWithCallback iterates through BPF map for each reason and direction.
```

carnerito replied:
@zacharysarah I'm really sorry for making confusion, but all changes you have proposed are not from my PR. I tried to amend the unsigned commit as suggested by @dylandreimerink and I've pushed someone else's commits to this branch by mistake.
The only thing that you have commented that is part of this PR is regarding the wording in metricsmap.go, and I have fixed that in [#d1988ae](https://github.com/cilium/cilium/pull/27370/commits/d1988ae903d5b7424af189c6bce16bf253c275d6)

zacharysarah replied:
@carnerito
> all changes you have proposed are not from my PR

~~And yet your changes touch them, creating an opportunity to make docs better. Otherwise--when will docs ever improve?~~

~~I'm disappointed in your choice to refuse uncomplicated improvements, but I'll approve regardless.~~

UPDATE: See https://github.com/cilium/cilium/pull/27370#discussion_r1296458751 and following 👇🏻

joestringer replied:
It looks to me like there was a misunderstanding here in the way that the changes got pushed into GitHub and the way that GitHub requested reviewers for these changes. From what I can tell, @carnerito never intended to make changes to these files in the PR, but the way that the PR was pushed into GitHub caused GitHub to pull @zacharysarah in for review.

In future it might be worth doing a quick double-check of the PR before marking it "ready for review" to make sure that GitHub presents the set of changes that you expected to present, so that GitHub only pulls in the codeowners for the areas that are being changed.

At this point it looks like @carnerito has been able to resolve this issue, maybe with a fresh push of the branch, and none of these comments are relevant to the changes in the PR any more. Either way it looks like this has been resolved in the end. @carnerito apologies for the confusion around this, thanks again for contributing these changes to Cilium :-)

zacharysarah replied:
@joestringer Thanks for the explanation. @carnerito Apologies for pushing back so strongly; I see the underlying issue now. Thanks for your contributions!

zacharysarah reviewed (changes requested):
@carnerito Thanks for the updates! Some minor stylistic changes are required, with strong encouragement to fix some of the inherited language touched by this PR.

christarazi reviewed (commented):
Thanks for the PR! Could you rebase and squash all commits together?

zacharysarah reviewed (approved):
LGTM

christarazi commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -137,47 +137,134 @@ func (vs Values) Bytes() uint64 {
+}
+
+func (mc *metricsmapCollector) Collect(ch chan<- prometheus.Metric) {
+
```
Nit: unnecessary newline

carnerito replied:
@christarazi Thanks for the review. I didn't know that maps have a BatchLookup method; I'll keep this in mind in the future. As for the newline, I was relying on gomt to do the formatting, but I completely agree that it would be better without it here.

christarazi reviewed (dismissed):
Very cool PR! I was wondering whether we could make use of batch lookups instead of using `IterateWithCallback`, but realized that the metrics map doesn't have very many entries in it, so maybe it's just a nice-to-have thing. With the PR there's already a performance improvement because the controller used to run every 5s and now it's based on the scrape interval which is very likely to be >=15s.

derailed commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -187,4 +274,14 @@ func init() {
+
+	// Register metrics map collector only once
+	once.Do(func() {
+		c := newMetricsMapCollector()
```
nit: no need for this tmp var. Call directly? `metrics.Register(newMetricsCollector())`

derailed commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -137,47 +137,134 @@ func (vs Values) Bytes() uint64 {
+func lookupOrCreateMetricsMapValues[K comparable](m map[K]metricValues, key K) metricValues {
+	if entry, ok := m[key]; ok {
+		return entry
+	} else {
```
nit: superfluous else

carnerito replied:
@derailed Thanks for review. I'll make changes after we resolve the issue with CI.

derailed reviewed (dismissed):
@carnerito Nice work Boris! Just a couple small picks...

carnerito committed:
Replace metricsmap-bpf-prom-sync with Prometheus Collector pattern

Previously, the 'metricsmap-bpf-prom-sync' controller was responsible for periodically collecting cilium_datapath drop
and forward metrics.
This commit introduces the metricsmapCollector, which implements the Prometheus Collector interface within the
metricsmap package. As a result, the aforementioned controller has been removed.
The metricsmapCollector is registered within the global Prometheus registry during the initialization of metricsmap.
The metrics.Register function has been modified to propagate errors from the registry.Register function instead of
simply overriding it with nil.
The metricsmapCollector comprises two metrics maps: forwardedMetricsMap and droppedMetricsMap.
These maps are populated within a callback function passed to the IterateWithCallback function.
This approach serves two primary purposes:
1. Separation of Map Iteration and Metric Update: By separating the iteration over the BPF map and the updating of
   Prometheus metrics, the implementation ensures that no partial metrics are exposed in case of map iteration failure.
2. Normalization of exposed Metrics: Unlike the statement in the bpf/lib/metrics.h comments, which suggests exposing
   only one reason label for forwarded metrics, the eBPF code exposes multiple reasons. Through testing, it was found
   that reasons 0 and 3 were exposed which triggered this error in prometheus client:
   https://github.com/prometheus/client_golang/issues/242

Furthermore, cilium command has indirect dependency on metricsmap package. To prevent metrics map initialization
to happen each time cilium command is executed, metricsmap is converted to hive Cell and injected in cilium daemon
Infrastructure module. For details see this comment: https://github.com/cilium/cilium/pull/27370#issuecomment-1710805663.

Fixes: #27058

Signed-off-by: Boris Petrovic <carnerito.b@gmail.com>

christarazi commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -187,4 +272,18 @@ func init() {
+					"cilium_datapath_drop/forward metrics will not be collected")
+			}
+		}()
+	})
```
Hm, the fact that this initialization occurs when a Cilium command is run first seems undesirable. Ideally, the daemon initializes all the metrics it needs upfront. The Cilium commands (which are separate binaries from the daemon) should not have to run any code to initialize metrics. 

I think what we could do here is to export `newMetricsMapCollector()` and have to be called within the `pkg/metrics/cell.go` code which is imported by the Agent Cell in `daemon/cmd/cells.go`.

cc @dylandreimerink, does this sound reasonable?

carnerito replied:
@dylandreimerink @christarazi Can you take a look [at latest commit](https://github.com/cilium/cilium/pull/27370/commits/54b1945191db724ac094c79e9785cf9f1b017d39). I have converted metricsmap to cell.Module which is invoked in `daemon/cmd/cells.go`. Now, only daemon will initialize metricsmap. I've tried to add it to `pkg/metrics/cell.go` as suggested, but than we have circular dependency issue
```
package github.com/cilium/cilium/api/v1/server
        imports github.com/cilium/cilium/pkg/metrics
        imports github.com/cilium/cilium/pkg/maps/metricsmap
        imports github.com/cilium/cilium/pkg/ebpf
        imports github.com/cilium/cilium/pkg/bpf
        imports github.com/cilium/cilium/pkg/controller
        imports github.com/cilium/cilium/pkg/metrics: import cycle not allowed
```
Furthermore, I've tried to add metricsmap collector to metricsmap module, like in Hive examples, with `cell.Metrics(newMetricsmapCollector)`, but I couldn't make it work that way. Even though all metricsmap metrics were registered, metricsmap Collector was not, so Collect method wasn't invoked on the scrape. I know that these metrics were registered because I've spawn separate goroutine that was incrementing counters periodically and I was able to scrape them. 
Is it possible to create custom Collector using metric types defined in `pkg/metrics/metric` package?

christarazi replied:
The structure looks good. I think it's unfortunate that there's an import cycle, which suggests that further refactoring is needed, but let's not tackle that in this PR. It would be good to keep that as a note in a code comment on the metricsmap cell. 

I'll let Dylan get to the finer details here.

christarazi reviewed (commented):
Thanks for the thorough investigation! 

Below is a comment on the metrics initialization code and regarding the bug in Cilium test helper functions, a separate PR would be good for that.

christarazi reviewed (commented):
I think you can squash all three commits into one.

derailed commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -137,54 +149,133 @@ func (vs Values) Bytes() uint64 {
+
+func (mc *metricsmapCollector) Collect(ch chan<- prometheus.Metric) {
+	err := Metrics.IterateWithCallback(func(key *Key, values *Values) {
+		if key.IsDrop() {
```
We could perhaps dry this up a bit by creating a metricsmap generic type and hang methods to upsert or even exporting metrics on the map vs using functions? i.e
```go
type metricsMap[k labels] map[k]metricValues

func (m metricsMap[labels]) upsert(l labels, b, c float64) {
	m[l] = m[l].inc(b, c)
}
// Then...
err := Metrics.IterateWithCallback(func(key *Key, values *Values) {
		if key.IsDrop() {
		  m.droppedMetricMap.upsert(dropLabels{...}, values.Bytes, value.Counts)
		} else {
		  m.forwardedMetricMap.upsert(forwardLabels{...}, values.Bytes, value.Counts)
		}
                ...
}
```

carnerito replied:
@derailed Thanks for review! 
I've created a generic `promMetrics` map with upsert() method as suggested. Code in a callback function looks bit cleaner now. 
Regarding exporting, I agree that it would be better if we could have `m.droppedMetricMap.export` and `m.forwardedMetricMap.export` instead those two for loops at the end of callback function, but in that case we would have to associate prometheus.Desc with labels and values. Since we have two values under same key (labels) in metrics map I don't see any elegant way to do this without rewriting most of the logic, if you have any hints, I'm willing to try :)

carnerito replied:
@derailed Hi, do you have any updates on this?

derailed replied:
@carnerito I think what you have now is indeed fine.

derailed reviewed (commented):
@carnerito Thank you Boris for this PR. Nice work!!

derailed reviewed (approved):
@carnerito Thank you for this PR and updates Boris!!
